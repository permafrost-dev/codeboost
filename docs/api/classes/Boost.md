[codeboost](../README.md) / [Exports](../modules.md) / Boost

# Class: Boost

## Table of contents

### Constructors

-   [constructor](Boost.md#constructor)

### Properties

-   [actions](Boost.md#actions)
-   [changedFiles](Boost.md#changedfiles)
-   [config](Boost.md#config)
-   [id](Boost.md#id)
-   [path](Boost.md#path)
-   [pullRequest](Boost.md#pullrequest)
-   [repositoryLimits](Boost.md#repositorylimits)
-   [runId](Boost.md#runid)
-   [scripts](Boost.md#scripts)
-   [state](Boost.md#state)
-   [version](Boost.md#version)

### Accessors

-   [appSettings](Boost.md#appsettings)
-   [history](Boost.md#history)

### Methods

-   [canRunOnRepository](Boost.md#canrunonrepository)
-   [checkoutPullRequestBranch](Boost.md#checkoutpullrequestbranch)
-   [createScriptHandlerParameters](Boost.md#createscripthandlerparameters)
-   [init](Boost.md#init)
-   [loadConfiguration](Boost.md#loadconfiguration)
-   [loadPullRequest](Boost.md#loadpullrequest)
-   [loadScripts](Boost.md#loadscripts)
-   [log](Boost.md#log)
-   [run](Boost.md#run)
-   [runInitializationScript](Boost.md#runinitializationscript)
-   [runScripts](Boost.md#runscripts)
-   [updatePullRequestBranchName](Boost.md#updatepullrequestbranchname)

## Constructors

### constructor

• **new Boost**(`codeBoost`, `boostPath`)

#### Parameters

| Name        | Type                        |
| :---------- | :-------------------------- |
| `codeBoost` | [`CodeBoost`](CodeBoost.md) |
| `boostPath` | `string`                    |

#### Defined in

[lib/Boost.ts:60](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L60)

## Properties

### actions

• **actions**: `any`[]

#### Defined in

[lib/Boost.ts:55](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L55)

---

### changedFiles

• **changedFiles**: `string`[] = `[]`

#### Defined in

[lib/Boost.ts:57](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L57)

---

### config

• **config**: [`BoostConfiguration`](../interfaces/BoostConfiguration.md)

#### Defined in

[lib/Boost.ts:41](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L41)

---

### id

• **id**: `string`

#### Defined in

[lib/Boost.ts:43](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L43)

---

### path

• **path**: `string`

#### Defined in

[lib/Boost.ts:42](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L42)

---

### pullRequest

• **pullRequest**: `Object`

#### Type declaration

| Name     | Type     |
| :------- | :------- |
| `body`   | `string` |
| `branch` | `string` |
| `title`  | `string` |

#### Defined in

[lib/Boost.ts:49](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L49)

---

### repositoryLimits

• **repositoryLimits**: `Object`

#### Type declaration

| Name                 | Type     |
| :------------------- | :------- |
| `maxRunsPerVersion`  | `number` |
| `minutesBetweenRuns` | `number` |

#### Defined in

[lib/Boost.ts:45](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L45)

---

### runId

• **runId**: `string`

#### Defined in

[lib/Boost.ts:58](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L58)

---

### scripts

• **scripts**: `any`[]

#### Defined in

[lib/Boost.ts:54](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L54)

---

### state

• **state**: `Record`<`string`, `any`\> = `{}`

#### Defined in

[lib/Boost.ts:56](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L56)

---

### version

• **version**: `string`

#### Defined in

[lib/Boost.ts:44](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L44)

## Accessors

### appSettings

• `get` **appSettings**(): [`AppSettings`](../interfaces/AppSettings.md)

#### Returns

[`AppSettings`](../interfaces/AppSettings.md)

#### Defined in

[lib/Boost.ts:81](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L81)

---

### history

• `get` **history**(): [`BoostHistory`](../modules.md#boosthistory)

#### Returns

[`BoostHistory`](../modules.md#boosthistory)

#### Defined in

[lib/Boost.ts:85](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L85)

## Methods

### canRunOnRepository

▸ **canRunOnRepository**(`repo`): `boolean`

#### Parameters

| Name   | Type                                      |
| :----- | :---------------------------------------- |
| `repo` | `string` \| [`Repository`](Repository.md) |

#### Returns

`boolean`

#### Defined in

[lib/Boost.ts:268](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L268)

---

### checkoutPullRequestBranch

▸ **checkoutPullRequestBranch**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Boost.ts:228](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L228)

---

### createScriptHandlerParameters

▸ **createScriptHandlerParameters**(`args`, `historyItem`): [`BoostScriptHandlerParameters`](../interfaces/BoostScriptHandlerParameters.md)

#### Parameters

| Name          | Type                                                    |
| :------------ | :------------------------------------------------------ |
| `args`        | `any`[]                                                 |
| `historyItem` | [`BoostHistoryItem`](../interfaces/BoostHistoryItem.md) |

#### Returns

[`BoostScriptHandlerParameters`](../interfaces/BoostScriptHandlerParameters.md)

#### Defined in

[lib/Boost.ts:205](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L205)

---

### init

▸ **init**(`boostPath`): `void`

#### Parameters

| Name        | Type     |
| :---------- | :------- |
| `boostPath` | `string` |

#### Returns

`void`

#### Defined in

[lib/Boost.ts:67](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L67)

---

### loadConfiguration

▸ **loadConfiguration**(`boostPath`): [`BoostConfiguration`](../interfaces/BoostConfiguration.md)

#### Parameters

| Name        | Type     |
| :---------- | :------- |
| `boostPath` | `string` |

#### Returns

[`BoostConfiguration`](../interfaces/BoostConfiguration.md)

#### Defined in

[lib/Boost.ts:93](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L93)

---

### loadPullRequest

▸ **loadPullRequest**(`pullRequest`): `Object`

#### Parameters

| Name          | Type                       |
| :------------ | :------------------------- |
| `pullRequest` | `Record`<`string`, `any`\> |

#### Returns

`Object`

| Name     | Type     |
| :------- | :------- |
| `body`   | `any`    |
| `branch` | `string` |
| `title`  | `any`    |

#### Defined in

[lib/Boost.ts:99](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L99)

---

### loadScripts

▸ **loadScripts**(`scripts`): `any`[]

#### Parameters

| Name      | Type       |
| :-------- | :--------- |
| `scripts` | `string`[] |

#### Returns

`any`[]

#### Defined in

[lib/Boost.ts:107](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L107)

---

### log

▸ **log**(`message`): `void`

#### Parameters

| Name      | Type  |
| :-------- | :---- |
| `message` | `any` |

#### Returns

`void`

#### Defined in

[lib/Boost.ts:89](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L89)

---

### run

▸ **run**(`repository`, `args?`): `Promise`<`undefined` \| `false`\>

#### Parameters

| Name         | Type                          | Default value |
| :----------- | :---------------------------- | :------------ |
| `repository` | [`Repository`](Repository.md) | `undefined`   |
| `args`       | `any`[]                       | `[]`          |

#### Returns

`Promise`<`undefined` \| `false`\>

#### Defined in

[lib/Boost.ts:123](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L123)

---

### runInitializationScript

▸ **runInitializationScript**(`params`): `Promise`<`void`\>

#### Parameters

| Name     | Type                                                                            |
| :------- | :------------------------------------------------------------------------------ |
| `params` | [`BoostScriptHandlerParameters`](../interfaces/BoostScriptHandlerParameters.md) |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Boost.ts:221](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L221)

---

### runScripts

▸ **runScripts**(`params`): `Promise`<`void`\>

#### Parameters

| Name     | Type                                                                            |
| :------- | :------------------------------------------------------------------------------ |
| `params` | [`BoostScriptHandlerParameters`](../interfaces/BoostScriptHandlerParameters.md) |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Boost.ts:257](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L257)

---

### updatePullRequestBranchName

▸ **updatePullRequestBranchName**(): `Promise`<`boolean`\>

update to a unique branch name if the branch already exists

#### Returns

`Promise`<`boolean`\>

#### Defined in

[lib/Boost.ts:235](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Boost.ts#L235)
