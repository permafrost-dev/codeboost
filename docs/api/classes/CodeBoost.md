[codeboost](../README.md) / [Exports](../modules.md) / CodeBoost

# Class: CodeBoost

## Hierarchy

-   [`HasLogger`](HasLogger.md)<`this`\>

    ↳ **`CodeBoost`**

## Table of contents

### Constructors

-   [constructor](CodeBoost.md#constructor)

### Properties

-   [appSettings](CodeBoost.md#appsettings)
-   [historyManager](CodeBoost.md#historymanager)
-   [logger](CodeBoost.md#logger)
-   [repositoryPrepared](CodeBoost.md#repositoryprepared)

### Methods

-   [createLogger](CodeBoost.md#createlogger)
-   [init](CodeBoost.md#init)
-   [log](CodeBoost.md#log)
-   [prepareRepository](CodeBoost.md#preparerepository)
-   [runBoost](CodeBoost.md#runboost)

## Constructors

### constructor

• **new CodeBoost**(`appSettings`, `historyManager`)

#### Parameters

| Name             | Type                                          |
| :--------------- | :-------------------------------------------- |
| `appSettings`    | [`AppSettings`](../interfaces/AppSettings.md) |
| `historyManager` | [`HistoryManager`](HistoryManager.md)         |

#### Overrides

[HasLogger](HasLogger.md).[constructor](HasLogger.md#constructor)

## Properties

### appSettings

• **appSettings**: [`AppSettings`](../interfaces/AppSettings.md)

---

### historyManager

• **historyManager**: [`HistoryManager`](HistoryManager.md)

---

### logger

• **logger**: `Logger`

#### Inherited from

[HasLogger](HasLogger.md).[logger](HasLogger.md#logger)

---

### repositoryPrepared

• **repositoryPrepared**: `boolean` = `false`

## Methods

### createLogger

▸ **createLogger**(`targets`, `defaultMeta`): `void`

#### Parameters

| Name          | Type                                     |
| :------------ | :--------------------------------------- |
| `targets`     | [`LogTarget`](../modules.md#logtarget)[] |
| `defaultMeta` | `Record`<`string`, `any`\>               |

#### Returns

`void`

#### Inherited from

[HasLogger](HasLogger.md).[createLogger](HasLogger.md#createlogger)

---

### init

▸ **init**(`repository`, `appSettings`): `Promise`<`void`\>

#### Parameters

| Name          | Type                                          |
| :------------ | :-------------------------------------------- |
| `repository`  | [`Repository`](Repository.md)                 |
| `appSettings` | [`AppSettings`](../interfaces/AppSettings.md) |

#### Returns

`Promise`<`void`\>

---

### log

▸ **log**(`message`, `meta?`): `void`

#### Parameters

| Name      | Type     | Default value |
| :-------- | :------- | :------------ |
| `message` | `string` | `undefined`   |
| `meta`    | `any`[]  | `[]`          |

#### Returns

`void`

#### Inherited from

[HasLogger](HasLogger.md).[log](HasLogger.md#log)

---

### prepareRepository

▸ **prepareRepository**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

---

### runBoost

▸ **runBoost**(`id`, `args`): `Promise`<[`Boost`](Boost.md)\>

#### Parameters

| Name   | Type       |
| :----- | :--------- |
| `id`   | `string`   |
| `args` | `string`[] |

#### Returns

`Promise`<[`Boost`](Boost.md)\>
