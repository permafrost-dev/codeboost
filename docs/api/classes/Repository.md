[codeboost](../README.md) / [Exports](../modules.md) / Repository

# Class: Repository

## Table of contents

### Constructors

-   [constructor](Repository.md#constructor)

### Properties

-   [name](Repository.md#name)
-   [owner](Repository.md#owner)
-   [path](Repository.md#path)

### Accessors

-   [git](Repository.md#git)
-   [info](Repository.md#info)

### Methods

-   [checkout](Repository.md#checkout)
-   [clone](Repository.md#clone)
-   [createFork](Repository.md#createfork)
-   [currentBranch](Repository.md#currentbranch)
-   [defaultBranch](Repository.md#defaultbranch)
-   [fullRepositoryName](Repository.md#fullrepositoryname)
-   [localBranches](Repository.md#localbranches)
-   [onBranch](Repository.md#onbranch)
-   [prepare](Repository.md#prepare)
-   [pushToFork](Repository.md#pushtofork)

## Constructors

### constructor

• **new Repository**(`fullRepositoryName`, `repositoryStoragePath`)

#### Parameters

| Name                    | Type     |
| :---------------------- | :------- |
| `fullRepositoryName`    | `string` |
| `repositoryStoragePath` | `string` |

#### Defined in

[lib/Repository.ts:34](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L34)

## Properties

### name

• **name**: `string`

#### Defined in

[lib/Repository.ts:10](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L10)

---

### owner

• **owner**: `string`

#### Defined in

[lib/Repository.ts:11](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L11)

---

### path

• **path**: `string`

#### Defined in

[lib/Repository.ts:12](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L12)

## Accessors

### git

• `get` **git**(): `SimpleGit`

#### Returns

`SimpleGit`

#### Defined in

[lib/Repository.ts:22](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L22)

---

### info

• `get` **info**(): [`RepositoryInfo`](../interfaces/RepositoryInfo.md)

#### Returns

[`RepositoryInfo`](../interfaces/RepositoryInfo.md)

#### Defined in

[lib/Repository.ts:15](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L15)

## Methods

### checkout

▸ **checkout**(`branchName`): `Promise`<`void`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:81](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L81)

---

### clone

▸ **clone**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:43](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L43)

---

### createFork

▸ **createFork**(): `Promise`<`void`\>

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:101](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L101)

---

### currentBranch

▸ **currentBranch**(): `Promise`<`string`\>

#### Returns

`Promise`<`string`\>

#### Defined in

[lib/Repository.ts:73](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L73)

---

### defaultBranch

▸ **defaultBranch**(): `Promise`<`string`\>

#### Returns

`Promise`<`string`\>

#### Defined in

[lib/Repository.ts:96](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L96)

---

### fullRepositoryName

▸ **fullRepositoryName**(): `string`

#### Returns

`string`

#### Defined in

[lib/Repository.ts:124](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L124)

---

### localBranches

▸ **localBranches**(): `Promise`<`BranchSummary`\>

#### Returns

`Promise`<`BranchSummary`\>

#### Defined in

[lib/Repository.ts:67](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L67)

---

### onBranch

▸ **onBranch**(`branchName`): `Promise`<`boolean`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`boolean`\>

#### Defined in

[lib/Repository.ts:77](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L77)

---

### prepare

▸ **prepare**(): `Promise`<`void`\>

Checks out the default branch and pulls down the latest changes

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:58](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L58)

---

### pushToFork

▸ **pushToFork**(`branchName`): `Promise`<`void`\>

#### Parameters

| Name         | Type     |
| :----------- | :------- |
| `branchName` | `string` |

#### Returns

`Promise`<`void`\>

#### Defined in

[lib/Repository.ts:120](https://github.com/permafrost-dev/codeboost/blob/0b270dd/src/lib/Repository.ts#L120)
