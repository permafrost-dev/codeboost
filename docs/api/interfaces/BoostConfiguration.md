[codeboost](../README.md) / [Exports](../modules.md) / BoostConfiguration

# Interface: BoostConfiguration

## Table of contents

### Properties

-   [id](BoostConfiguration.md#id)
-   [pull_request](BoostConfiguration.md#pull_request)
-   [repository_limits](BoostConfiguration.md#repository_limits)
-   [scripts](BoostConfiguration.md#scripts)
-   [version](BoostConfiguration.md#version)

## Properties

### id

• **id**: `string`

---

### pull_request

• **pull_request**: `Object`

#### Type declaration

| Name     | Type     |
| :------- | :------- |
| `body`   | `string` |
| `branch` | `string` |
| `title`  | `string` |

---

### repository_limits

• **repository_limits**: `Object`

#### Type declaration

| Name                   | Type     |
| :--------------------- | :------- |
| `max_runs_per_version` | `number` |
| `minutes_between_runs` | `number` |

---

### scripts

• **scripts**: `Object`

#### Type declaration

| Name       | Type       |
| :--------- | :--------- |
| `files`    | `string`[] |
| `parallel` | `boolean`  |

---

### version

• **version**: `string`
