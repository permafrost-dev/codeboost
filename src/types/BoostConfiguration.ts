export interface BoostConfiguration {
    id: string;
    version: string;
    repository_limits: {
        max_runs_per_version: number;
        minutes_between_runs: number;
    };
    pull_request: {
        title: string;
        body: string;
        branch: string;
    };
    scripts: {
        parallel: boolean;
        files: string[];
    };
    actions: any[];
}
