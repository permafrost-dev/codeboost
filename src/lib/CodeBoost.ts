import { AppSettings } from '@/lib/AppSettings';
import { Boost } from '@/lib/Boost';
import { HasLogger, LogTarget } from '@/traits/HasLogger';
import { HistoryManager } from '@/lib/HistoryManager';
import { Repository } from '@/lib/Repository';
import { Mixin } from 'ts-mixer';

export class CodeBoost extends Mixin(HasLogger) {
    protected repository!: Repository;
    public appSettings!: AppSettings;
    public historyManager!: HistoryManager;
    public repositoryPrepared = false;

    constructor(appSettings: AppSettings, historyManager: HistoryManager) {
        super();
        this.createLogger(<LogTarget[]>appSettings?.log_target ?? [], {});
        this.appSettings = appSettings;
        this.historyManager = historyManager;
    }

    public async init(repository: Repository, appSettings: AppSettings) {
        this.appSettings = appSettings;
        this.repository = repository;
    }

    public async prepareRepository() {
        if (this.repositoryPrepared) {
            return;
        }

        await this.repository?.clone();
        await this.repository?.prepare();

        if (this.appSettings.use_forks && !this.appSettings.dry_run) {
            console.log('creating fork');
            await this.repository?.createFork();
        }

        this.repositoryPrepared = true;
    }

    public async runBoost(boost: string | Boost, args: string[]) {
        boost = typeof boost === 'string' ? this.getBoost(boost) : boost;

        await boost.run(this.repository, args);

        return boost;
    }

    public getBoost(boostName: string) {
        return new Boost(this, `${this.appSettings.boosts_path}/${boostName}`);
    }
}
