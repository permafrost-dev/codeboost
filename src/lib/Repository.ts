import { createOctokit, Github } from '@/lib/github';
import { parseFullRepositoryName } from '@/lib/stringHelpers';
import { execSync } from 'child_process';
import { existsSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import simpleGit, { SimpleGit } from 'simple-git';

export class Repository {
    public name: string;
    public owner: string;
    public path: string;
    protected gitInstance!: SimpleGit;

    public get git() {
        if (this.gitInstance === undefined) {
            this.gitInstance = simpleGit({
                baseDir: this.path,
                binary: 'git',
                maxConcurrentProcesses: 2,
                trimmed: false,
            });
        }
        return this.gitInstance;
    }

    constructor(fullRepositoryName: string, repositoryStoragePath: string) {
        const { owner, name } = parseFullRepositoryName(fullRepositoryName);

        this.name = name;
        this.owner = owner;

        this.path = `${repositoryStoragePath}/${fullRepositoryName}`;
    }

    public async clone() {
        const parentDir = dirname(this.path);

        if (!existsSync(parentDir)) {
            mkdirSync(parentDir, { recursive: true });
        }

        if (!existsSync(this.path)) {
            execSync(`git -C '${parentDir}' clone git@github.com:${this.owner}/${this.name}.git`, { stdio: 'pipe' });
        }
    }

    /**
     * Checks out the default branch and pulls down the latest changes
     */
    public async prepare() {
        await this.checkout(await this.defaultBranch());
        await this.git.pull('origin');
    }

    public async currentBranch() {
        return (await this.git.branchLocal()).current;
    }

    public async onBranch(branchName: string) {
        return (await this.currentBranch()) === branchName;
    }

    public async checkout(branchName: string) {
        if (!(await this.onBranch(branchName))) {
            await this.git.checkoutLocalBranch(branchName);
        }
    }

    public async defaultBranch() {
        const result = await this.git.revparse('--abbrev-ref', [ 'origin/HEAD' ]);
        return result.replace(/^.+\//, '');
    }

    public async createFork() {
        const octokit = createOctokit();

        try {
            const result = await Github.forkRepository(this, octokit);
            this.git.addRemote('fork', result.ssh_url);
            return;
        } catch (e) {
            console.error(e);
        }

        try {
            const result = await Github.getRepository(this, octokit);
            this.git.addRemote('fork', result.ssh_url);
        } catch (e) {
            console.error(e);
        }
    }

    async pushToFork(branchName: string) {
        await this.git.push('fork', branchName);
    }

    fullRepositoryName() {
        return `${this.owner}/${this.name}`;
    }
}
