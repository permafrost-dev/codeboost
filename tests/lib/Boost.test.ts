/* eslint-disable sort-keys */

import { Boost } from '@/lib/Boost';
import { CodeBoost } from '@/lib/CodeBoost';
import { Repository } from '@/lib/Repository';
import { BoostConfiguration } from '@/types/BoostConfiguration';
import { FakeHistoryManager } from '@tests/fakes/FakeHistoryManager';

const createBoost = (path, config = {}, historyMgr: any = null, codeboost: any = null) => {
    historyMgr = historyMgr ?? new FakeHistoryManager();
    codeboost = codeboost ?? new CodeBoost(historyMgr);

    const result = new Boost(codeboost, path);
    result.config = Object.assign({}, result.config, config);
    result.init(path);

    result.path = result.path.replace(__dirname, '.');

    return result;
};

const createBoostMock = methodName => jest.spyOn(Boost.prototype, methodName);

it('sets the correct properties on create', async () => {
    const boostConfig: BoostConfiguration = {
        id: 'test-one',
        version: '1.0.0',
        repository_limits: {
            max_runs_per_version: 1,
            minutes_between_runs: 60,
        },
        pull_request: {
            title: 'Add Test One',
            body: 'This PR is a test',
            branch: 'add-test-one',
        },
        scripts: {
            parallel: true,
            files: [],
        },
        actions: [],
    };

    const boost = new Boost(new CodeBoost(new FakeHistoryManager()), `${__dirname}/../fixtures/test-boost-1`);
    boost.config = boostConfig;
    boost.init(`${__dirname}/../fixtures/test-boost-1`);

    const keys = Object.keys(boost);
    keys.sort();

    expect(keys).toMatchSnapshot();
});

it('loads a boost from a path', async () => {
    const config = { repository_limits: { max_runs_per_version: 1, minutes_between_runs: 5 } };
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, new FakeHistoryManager());

    expect(boost.scripts[0]).toBeInstanceOf(Function);

    const keys = Object.keys(boost);
    keys.sort();

    expect(keys).toMatchSnapshot();
});

it('respects the max runs per version limit', async () => {
    const config = { repository_limits: { max_runs_per_version: 1, minutes_between_runs: 5 } };
    const historyMgr = new FakeHistoryManager();
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, historyMgr);

    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, '0.0.0', 'owner1/name1');
    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner1/name1');
    expect(boost.canRunOnRepository('owner1/name1')).toBeFalsy();
});

it('respects the min minutes between runs limit', async () => {
    const config = { repository_limits: { max_runs_per_version: 999, minutes_between_runs: 5 } };
    const historyMgr = new FakeHistoryManager();
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, historyMgr);

    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner1/name1', 10);
    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner1/name1', 3);
    expect(boost.canRunOnRepository('owner1/name1')).toBeFalsy();
});

it('does not consider skipped runs when determining if limits apply', async () => {
    const config = { repository_limits: { max_runs_per_version: 1, minutes_between_runs: 5 } };
    const historyMgr = new FakeHistoryManager();
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, historyMgr);

    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSkippedItem(boost.id, boost.version, 'owner1/name1', 3);
    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner1/name1', 3);
    expect(boost.canRunOnRepository('owner1/name1')).toBeFalsy();
});

it('does not consider runs for other repositories when determining if limits apply', async () => {
    const config = { repository_limits: { max_runs_per_version: 1, minutes_between_runs: 5 } };
    const historyMgr = new FakeHistoryManager();
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, historyMgr);

    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner2/name2', 3);
    expect(boost.canRunOnRepository('owner1/name1')).toBeTruthy();
    historyMgr.addSucceededItem(boost.id, boost.version, 'owner1/name1', 3);
    expect(boost.canRunOnRepository('owner1/name1')).toBeFalsy();
});

it('mocks a class method', () => {
    const runScriptsMock = jest.spyOn(Boost.prototype, 'runScripts').mockImplementation(async () => {});

    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`);
    boost.runScripts(<any>{});

    expect(runScriptsMock).toHaveBeenCalledTimes(1);

    runScriptsMock.mockRestore();
});

it(`runs all of a boost's scripts synchronously`, async () => {
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, { scripts: { parallel: false, files: [] } });
    boost.scripts = [ jest.fn(), jest.fn() ];

    await boost.runScripts(<any>{});

    boost.scripts.forEach(script => expect(script).toHaveBeenCalledTimes(1));
});

it(`runs all of a boost's scripts asynchronously`, async () => {
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, { scripts: { parallel: true, files: [] } });
    boost.scripts = [ jest.fn(), jest.fn() ];

    await boost.runScripts(<any>{});

    boost.scripts.forEach(script => expect(script).toHaveBeenCalledTimes(1));
});

it('runs on a repository', async () => {
    const checkoutPullBranchMock = createBoostMock('checkoutPullRequestBranch').mockImplementation(async () => {});
    const createScriptHandlerParametersMock = createBoostMock('createScriptHandlerParameters').mockImplementation(() => {
        return <any>{};
    });
    const mocks = [ checkoutPullBranchMock, createScriptHandlerParametersMock ];

    const codeboost = new CodeBoost(new FakeHistoryManager());
    codeboost.appSettings = { use_pull_requests: true } as any;

    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, {}, codeboost.historyManager, codeboost);
    boost.scripts = [];

    const repo = new Repository('owner1/name1', `${__dirname}/../fixtures/repos`);

    await boost.run(repo, []);

    mocks.forEach(mock => expect(mock).toHaveBeenCalledTimes(1));

    mocks.forEach(mock => mock.mockRestore());
});

it('runs on a repository and creates a history item', async () => {
    const checkoutPullBranchMock = createBoostMock('checkoutPullRequestBranch').mockImplementation(async () => {});
    const createScriptHandlerParametersMock = createBoostMock('createScriptHandlerParameters').mockImplementation(() => {
        return <any>{};
    });
    const mocks = [ checkoutPullBranchMock, createScriptHandlerParametersMock ];

    const codeboost = new CodeBoost(new FakeHistoryManager());
    codeboost.appSettings = { use_pull_requests: true } as any;

    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, {}, codeboost.historyManager, codeboost);
    const repo = new Repository('owner1/name1', `${__dirname}/../fixtures/repos`);

    boost.scripts = [];

    await boost.run(repo, []);

    expect(codeboost.historyManager.data).toHaveLength(1);
    expect(codeboost.historyManager.data[0].state).not.toBe('running');
    expect(codeboost.historyManager.data[0].finished_at).not.toBeNull();

    mocks.forEach(m => m.mockRestore());
});

it('runs on a repository and creates a skipped history item if limits apply', async () => {
    const codeboost = new CodeBoost(new FakeHistoryManager());
    codeboost.appSettings = { use_pull_requests: true } as any;
    const config = { repository_limits: { max_runs_per_version: 1, minutes_between_runs: 60 } };
    const boost = createBoost(`${__dirname}/../fixtures/test-boost-1`, config, codeboost.historyManager, codeboost);
    const repo = new Repository('owner1/name1', `${__dirname}/../fixtures/repos`);

    (codeboost.historyManager as FakeHistoryManager).addSucceededItem(boost.id, boost.version, repo.fullRepositoryName(), 3);

    await boost.run(repo, []);

    expect(codeboost.historyManager.data).toHaveLength(2);
    expect(codeboost.historyManager.data[1].state).toBe('skipped');
});
