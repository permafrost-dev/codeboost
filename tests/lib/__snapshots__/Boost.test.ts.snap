// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads a boost from a path 1`] = `
[
  "codeBoost",
  "config",
  "appSettings",
  "repository",
  "path",
  "id",
  "version",
  "repositoryLimits",
  "pullRequest",
  "scripts",
  "actions",
  "state",
  "changedFiles",
  "runId",
]
`;

exports[`sets the correct properties on create 1`] = `
[
  "codeBoost",
  "config",
  "appSettings",
  "repository",
  "path",
  "id",
  "version",
  "repositoryLimits",
  "pullRequest",
  "scripts",
  "actions",
  "state",
  "changedFiles",
  "runId",
]
`;
